crates/data_components/src/s3_vectors/put_vectors_sink.rs

```diff
@@ -0,0 +1,442 @@
+    #[must_use]
+    pub fn new(table: S3VectorsTable) -> Self {
+        Self { table }
+    }
+}
+
+impl std::fmt::Debug for PutVectorsSink {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PutVectorsExec").finish_non_exhaustive()
+    }
+}
+
+impl DisplayAs for PutVectorsSink {
+    fn fmt_as(&self, _t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "PutVectorsExec")
+    }
+}
+
+#[async_trait]
+impl DataSink for PutVectorsSink {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn schema(&self) -> &SchemaRef {
+        &self.table.schema
+    }
+
+    async fn write_all(
+        &self,
+        mut data: SendableRecordBatchStream,
+        _context: &Arc<TaskContext>,
+    ) -> DataFusionResult<u64> {
+        let mut count = 0;
+
+        while let Some(record_batch) = data.next().await {
+            let record_batch = record_batch?;
+
+            let vectors = create_put_input_vectors(&record_batch)?;
+
+            let (index_arn, vector_bucket_name, index_name) =
+                self.table.idx.index_identifier_variables();
+
+            for chunk in vectors.chunks(PUT_VECTORS_MAX_ITEMS) {
+                self.table
+                    .client
+                    .put_vectors(
+                        PutVectorsInput::builder()
+                            .set_index_arn(index_arn.clone())
+                            .set_index_name(index_name.clone())
+                            .set_vector_bucket_name(vector_bucket_name.clone())
+                            .set_vectors(Some(chunk.to_vec()))
+                            .build()
+                            .context(BuildInputSnafu)?,
+                    )
+                    .await
+                    .map_err(SdkError::into_service_error)
+                    .boxed()
+                    .context(PutVectorsSnafu)?;
+
+                count += chunk.len();
+            }
+        }
+
+        Ok(count as _)
+    }
+}
+
+fn create_put_input_vectors(record_batch: &RecordBatch) -> Result<Vec<PutInputVector>> {
+    let name = S3_VECTOR_PRIMARY_KEY_NAME.to_string();
+    let keys = record_batch
+        .column_by_name(&name)
+        .ok_or_else(|| Error::MissingColumn { name: name.clone() })?
+        .as_any()
+        .downcast_ref::<arrow::array::StringArray>()
+        .ok_or_else(|| Error::ColumnTypeMismatch {
+            name,
+            expected: "StringArray".to_string(),
+        })?;
+
+    let name = S3_VECTOR_EMBEDDING_NAME.to_string();
+    let vectors = record_batch
+        .column_by_name(&name)
+        .ok_or_else(|| Error::MissingColumn { name: name.clone() })?
+        .as_any()
+        .downcast_ref::<arrow::array::ListArray>()
+        .ok_or_else(|| Error::ColumnTypeMismatch {
+            name,
+            expected: "ListArray".to_string(),
+        })?;
+
+    let schema = record_batch.schema();
+    let fields = schema
```

Comment:
It boils down to the same logic, but you could tag columns in the schema as metadata only and then collect those (if you wanted to decouple column name assertions in this specific stage)